dvc/repo/index.py

```diff
@@ -113,6 +114,62 @@ def is_stage_inside_path(stage: "Stage") -> bool:
 
         return self.filter(is_stage_inside_path)
 
+    @staticmethod
+    def _hash_targets(
```

Comment:
Hm, what is this for?

Reply:
Caching known sets of targets that we've already collected using this index. It's just the python `__hash__()` of a frozen set of stages for a given `targets` string + deps/recursive collection flags.

So we can do the
```
targets_hash = self._hash_targets(targets, **kwargs)
if targets_hash not in self._collected_targets:
```

Maybe it would be easier to read code-wise if it was
```
targets = (frozenset(targets), with_deps, recursive)
collected_targets[targets] = ...
```
but we don't really need to re-use the entire frozenset from the key at all so I simplified it here to just use the __hash__